File: 프로그래머스/2/42860. 조이스틱/조이스틱.swift
// A

// 좌우 - 커서 이동
// 위아래 - 알파벳 변경

// 최소 조작횟수

// 대문자로만, Name은 1~20

// 좌 -> 우
// 위 아래로 변경하면 되는거 아닌가?

// 아스키코드 값 차이

import Foundation

func solution(_ name:String) -> Int {
    
    let n = name.count
    let arr = Array(name)
    
    // 위 아래 이동
    var upDown = 0
    for char in arr {
        let diff = abs(Int(char.asciiValue!) - Int(Character("A").asciiValue!))
        upDown += min(diff, abs(26 - diff))
    }
    
    // 좌우 이동
    var leftRight = n - 1
    for i in 0..<n {
        
        var next = i + 1
        while next < n && arr[next] == "A" {
            next += 1
        }
        
        let left = i - 0
        let right = n - next
        
        leftRight = min(leftRight, left + right + min(left, right))
    }
    
    return upDown + leftRight
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
# 조이스틱 알파벳 이름
# 맨 처음만 A

# 위 - 다음
# 아 - 이전
# 왼 - 커서
# 오 - 커서 

# AAAA 로 구성된 이름
# 굴려서 일치하는 이름으로 만들어라

# 움직이는 최솟값

# 다 연결되어 있음 > 최소 좌우 이동, 최소 상하 이동의 합은 몇인가

# 좌우 이동 > 
# 상하 이동 > 아스키 코드 차이 x: min(26 - x, x)

def solution(name):
    
    n = len(name)
    cur = ['A'] * n
    s = 0
    e = 0
    
    for i in range(n):
        if cur[i] != name[i]:
            s = i
            break
            
    for i in range(n-1, -1, -1):
        if cur[i] != name[i]:
            e = i
            break
    
    h_mv = min(e, n-s)
    
    target = []
    for i in range(n):
        if name[i] != 'A':
            target.append(i)
    
    for i in range(len(target)-1):
        s = target[i]
        e = target[i+1]
        
        if s < n-e:
            h_mv = min(h_mv, s + s + n-e)        
        else:
            h_mv = min(h_mv, n-e + n-e + s)     
        
    v_mv = 0
    for i in range(n):
        if cur[i] != name[i]:
            v_mv += mv_updown(cur[i], name[i])
    
    return h_mv + v_mv

def mv_updown(a, b):
    x = abs(ord(a) - ord(b))
    return min(26-x, x)

# 오른쪽으로 가는 경우 